### PR TITLE
feat(sampling): add sampling for all breakdowns

### PR DESCRIPTION
**Pull request by yakkomajuri**
Putting this up as a WIP as a way of having a clearer think about everything e.g. should we allow sampling on all these breakdown queries?

### CODE REVIEW

**Comment by neilkakkar** [comment]
I think breakdowns don't face the same problem as retention (even if both kinds have 2 different samplings in the same query) because breakdowns are sampling in aggregate, i.e. they use the sampling to generate possible breakdown keys (of which the smaller counts are discarded anyways until you press 'load more breakdown values').

So, in that sense, making this faster is _ideal_, as it doesn't affect the final result. Maybe we can even sample these even when other things are not sampled (as long as event volume is high enough). They have fewer, easily traded-off gotchas when compared to the distinctID sampling of a trend line/funnel.

The only point of concern here is stability when there are very few breakdown values: Clicking on load more shouldn't duplicate values, nor miss them; but in that case we shouldn't need sampling anyway.

File: frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx [hunk 1]
```diff
@@ -17,6 +17,8 @@ const INSIGHT_TYPES_WITH_SAMPLING_SUPPORT = new Set([
     InsightType.RETENTION,
 ])
 
+const INSIGHT_TYPES_WITH_SAMPLING_SUPPORT = new Set([InsightType.LIFECYCLE, InsightType.FUNNELS, InsightType.TRENDS])
+
```
Comment (yakkomajuri) [inline frontend/src/scenes/insights/EditorFilters/SamplingFilter.tsx:21]:
```suggestion
```

**Comment by pauldambra** [comment]
Still too WIP to test locally?

File: posthog/queries/breakdown_props.py [hunk 2]
```diff
@@ -147,6 +147,8 @@ def get_breakdown_prop_values(
         cast_as_float=filter.using_histogram,
     )
 
+    sample_clause = f"SAMPLE {filter.sampling_factor}" if filter.sampling_factor else ""
```
Comment (macobo) [inline posthog/queries/breakdown_props.py:150]: Interpolation is kind of :( as a practice but checked the code and it seems defensible here.

File: posthog/queries/util.py [hunk 3]
```diff
@@ -103,7 +103,9 @@ def convert_to_datetime_aware(date_obj):
     return date_obj
 
 
-def correct_result_for_sampling(value: int, sampling_factor: Optional[float], entity_math: Optional[str] = None) -> int:
+def correct_result_for_sampling(
+    value: Union[int, float], sampling_factor: Optional[float], entity_math: Optional[str] = None
+) -> int:
```
Comment (macobo) [inline posthog/queries/util.py:108]: What's going on with the typing here? Should we retain typing here instead of casting-to-int if dealing with floats?

Comment (yakkomajuri) [reply posthog/queries/util.py:108]: yeah I made a mess here fighting with CI - will retain types

**Comment by yakkomajuri** [comment]
> Still too WIP to test locally?

hmm unsure what's happened there - shouldn't be the case

File: posthog/queries/breakdown_props.py (hunk 2 above)
Comment (yakkomajuri) [reply posthog/queries/breakdown_props.py:150]: yeah I might work on a refactor once everything is in to do params everywhere

**Comment by yakkomajuri** [comment]
@pauldambra could you inspect the CH queries and send those over so I can take a look? (cmd+k --> Debug ClickHouse queries but you probably know this)

**Comment by pauldambra** [comment]
Looks like this one 

```
/* user_id:1 request:_api_projects_1_insights_trend_ */ 
SELECT groupArray(value) FROM (
    SELECT
        replaceRegexpAll(JSONExtractRaw(properties, '$browser'), '^"|"$', '') AS value,
        count(*) as count
    FROM events e
    SAMPLE 0.1



    WHERE
        team_id = 1 AND event = '$pageview' AND toTimeZone(timestamp, 'UTC') >= toDateTime('2023-02-28 00:00:00', 'UTC') AND toTimeZone(timestamp, 'UTC') <= toDateTime('2023-03-07 23:59:59', 'UTC')
    GROUP BY value
    ORDER BY count DESC, value DESC
    LIMIT 25 OFFSET 0
)
```

With sample on I get an empty array, without I get some results. Presumably there aren't enough rows locally for the fixed sampling to return data. So, a dev issue rather than a prod issue. But confusing in future for folk... 

Same thing I hit with sampling the heatmap, locally I got no results so went with `SAMPLE N`. 

Maybe we do need a "you don't have enough data to bother sampling" check... Not blocking for this PR though!

**Comment by yakkomajuri** [comment]
Yeah I'll merge and check the data in prod and see!

**Comment by yakkomajuri** [comment]
all good in prod @pauldambra but let me know if you see any issues!

**Pull request merged**